mongosql/src/algebrizer/definitions.rs

```diff
@@ -904,30 +998,126 @@ impl<'a> Algebrizer<'a> {
+                }
+            },
+        }
+    }
+
+    fn build_distinct_star_group_and_project_stage(
+        &self,
+        source: mir::Stage,
+    ) -> Result<mir::Stage> {
+        let source_result_set = source.schema(&self.schema_inference_state())?;
+        let mut bottom = None;
+        let mut datasource_names = Vec::new();
+
+        // Identify data sources from the result set schema
+        for key in source_result_set.schema_env.keys() {
+            match &key.datasource {
+                DatasourceName::Named(n) => {
+                    datasource_names.push((n.clone(), key.clone()));
+                }
+                DatasourceName::Bottom => {
+                    bottom = Some(Key::bot(self.scope_level));
+                }
             }
         }
+
+        let mut group_keys: Vec<OptionallyAliasedExpr> = datasource_names
+            .iter()
+            .map(|(datasource_name, key)| {
+                OptionallyAliasedExpr::Aliased(AliasedExpr {
+                    alias: datasource_name.clone(),
+                    expr: Expression::Reference(ReferenceExpr { key: key.clone() }),
+                })
+            })
+            .collect();
+
+        if let Some(bottom_key) = &bottom {
+            group_keys.push(OptionallyAliasedExpr::Aliased(AliasedExpr {
+                alias: BOTTOM_ALIAS.to_string(),
+                expr: Expression::Reference(ReferenceExpr {
+                    key: bottom_key.clone(),
+                }),
+            }));
+        }
+
+        // Adding the group stage to deduplicate rows
+        let group_stage = mir::Stage::Group(mir::Group {
+            source: Box::new(source),
+            keys: group_keys,
+            aggregations: vec![],
+            cache: SchemaCache::new(),
+            scope: self.scope_level,
+        });
+
+        let mut expression = BindingTuple::new();
+        for (datasource_name, _) in &datasource_names {
+            expression.insert(
+                Key::named(datasource_name, self.scope_level),
+                Expression::FieldAccess(FieldAccess {
+                    expr: Box::new(Expression::Reference(ReferenceExpr {
+                        key: Key::bot(self.scope_level),
+                    })),
+                    field: datasource_name.clone().to_string(),
+                    is_nullable: false,
+                }),
+            );
+        }
+
+        if bottom.is_some() {
+            expression.insert(
+                Key {
+                    datasource: DatasourceName::Bottom,
+                    scope: self.scope_level,
+                },
+                Expression::FieldAccess(FieldAccess {
+                    expr: Box::new(Expression::Reference(ReferenceExpr {
+                        key: Key::bot(self.scope_level),
+                    })),
+                    field: BOTTOM_ALIAS.to_string(),
+                    is_nullable: false,
+                }),
+            );
+        }
```

Comment:
```suggestion
        for (group_key_name, key) in &datasources {
            expression.insert(
                key.clone(), // given the way naming works, we need to use the same key!
                Expression::FieldAccess(FieldAccess {
                    expr: Box::new(Expression::Reference(ReferenceExpr {
                        key: Key::bot(self.scope_level),
                    })),
                    field: group_key_name.clone().to_string(),
                    is_nullable: false,
                }),
            );
        }
```
Consequent change based on the proposal above.